core/trino-main/src/main/java/io/trino/sql/planner/iterative/rule/ImplementTableFunctionSource.java

```diff
@@ -0,0 +1,758 @@
+                Assignments.builder()
+                        .putIdentities(joinedNodes.joinedNode().getOutputSymbols())
+                        .put(joinedRowNumber, rowNumberExpression)
+                        .put(joinedPartitionSize, partitionSizeExpression)
+                        .build());
+
+        List<Symbol> joinedPartitionBy = ImmutableList.<Symbol>builder()
+                .addAll(joinedNodes.leftPartitionBy())
+                .addAll(joinedNodes.rightPartitionBy())
+                .build();
+
+        boolean joinedPruneWhenEmpty = joinedNodes.leftPruneWhenEmpty() || joinedNodes.rightPruneWhenEmpty();
+
+        Map<Symbol, Symbol> joinedRowNumberSymbolsMapping = ImmutableMap.<Symbol, Symbol>builder()
+                .putAll(joinedNodes.leftRowNumberSymbolsMapping())
+                .putAll(joinedNodes.rightRowNumberSymbolsMapping())
+                .buildOrThrow();
+
+        return new NodeWithSymbols(project, joinedRowNumber, joinedPartitionSize, joinedPartitionBy, joinedPruneWhenEmpty, joinedRowNumberSymbolsMapping);
+    }
+
+    private static NodeWithMarkers appendMarkerSymbols(PlanNode node, Set<Symbol> symbols, Symbol referenceSymbol, Context context)
+    {
+        Assignments.Builder assignments = Assignments.builder();
+        assignments.putIdentities(node.getOutputSymbols());
+
+        ImmutableMap.Builder<Symbol, Symbol> symbolsToMarkers = ImmutableMap.builder();
+
+        for (Symbol symbol : symbols) {
+            Symbol marker = context.getSymbolAllocator().newSymbol("marker", BIGINT);
+            symbolsToMarkers.put(symbol, marker);
+            Expression actual = symbol.toSymbolReference();
+            Expression reference = referenceSymbol.toSymbolReference();
+            assignments.put(marker, new IfExpression(new ComparisonExpression(EQUAL, actual, reference), actual, new Cast(new NullLiteral(), toSqlType(BIGINT))));
+        }
+
+        PlanNode project = new ProjectNode(
+                context.getIdAllocator().getNextId(),
+                node,
+                assignments.build());
+
+        return new NodeWithMarkers(project, symbolsToMarkers.buildOrThrow());
+    }
+
+    private record SourceWithProperties(PlanNode source, TableArgumentProperties properties)
+    {
+        SourceWithProperties
+        {
+            requireNonNull(source, "source is null");
+            requireNonNull(properties, "properties is null");
+        }
+    }
+
+    private record NodeWithSymbols(PlanNode node, Symbol rowNumber, Symbol partitionSize, List<Symbol> partitionBy, boolean pruneWhenEmpty, Map<Symbol, Symbol> rowNumberSymbolsMapping)
+    {
+        NodeWithSymbols
+        {
+            requireNonNull(node, "node is null");
+            requireNonNull(rowNumber, "rowNumber is null");
+            requireNonNull(partitionSize, "partitionSize is null");
+            partitionBy = ImmutableList.copyOf(requireNonNull(partitionBy, "partitionBy is null"));
```

Comment:
requireNonNull not needed for partitionBy or rowNumberSymbolsMapping since they will cause the copyOf to fail anyway